### DISCUSSION TITLE
upgrade to netbox 3.10 - failure to connect to postgreSQL

### DISCUSSION THREAD

**Discussion by jcralbino** [netbox-community/netbox#7889] [General]
Hello Team,

I have seen an issue during an upgrade process from 3.0.9 to 3.0.10. 

As in 3.0.10 the psycopg2-binary v2.9.2 is introduced an error connecting to the postgreSQL

Downgrading this to psycopg2-binary v2.9.1 solved the issue in the 
[release details](https://www.psycopg.org/docs/news.html#current-release )
i see this:

- Wheel package compiled against OpenSSL 1.1.1l and PostgreSQL 14.1 (ticket #1388).

Has this been an issue seen by other also?

**Comment by candlerb** [comment]
I am running netbox 3.0.10, on Ubuntu 20.04, python 3.8, openssl 1.1.1f, and with postgres 12.9 (all standard out-of-box)

```
root@netbox3:~# cd /opt/netbox
root@netbox3:/opt/netbox# . venv/bin/activate
(venv) root@netbox3:/opt/netbox# pip list | grep psyco
psycopg2-binary               2.9.2
(venv) root@netbox3:/opt/netbox#
```

I don't see any problem with this version of psycopg2.

What is it that you do that causes the problem?  What exact error do you see?  What OS are you on?

**Reply by amk1969** [reply]
Hi, I am the one who actually performed the install, so let me complement the report...
The OS is RHEL 8.4, python 3.9 and Postgres client ver 13.3, all from official RedHat repositories. (not sure why Netbox install guide claims that _CentOS 8 does not provide Python 3.7 or later via its native package manager._  see https://pkgs.org/search/?q=python39 for example)
The postgres server is remote, same OS and version.

The error looks like this:
```
Applying database migrations (python3 netbox/manage.py migrate)...
Traceback (most recent call last):
  File "/opt/netbox-3.0.10/venv/lib64/python3.9/site-packages/django/db/backends/base/base.py", line 219, in ensure_connection
    self.connect()
  File "/opt/netbox-3.0.10/venv/lib64/python3.9/site-packages/django/utils/asyncio.py", line 33, in inner
    return func(*args, **kwargs)
  File "/opt/netbox-3.0.10/venv/lib64/python3.9/site-packages/django/db/backends/base/base.py", line 200, in connect
    self.connection = self.get_new_connection(conn_params)
  File "/opt/netbox-3.0.10/venv/lib64/python3.9/site-packages/django/utils/asyncio.py", line 33, in inner
    return func(*args, **kwargs)
  File "/opt/netbox-3.0.10/venv/lib64/python3.9/site-packages/django/db/backends/postgresql/base.py", line 187, in get_new_connection
    connection = Database.connect(**conn_params)
  File "/opt/netbox-3.0.10/venv/lib64/python3.9/site-packages/psycopg2/__init__.py", line 122, in connect
    conn = _connect(dsn, connection_factory=connection_factory, **kwasync)
psycopg2.OperationalError: connection to server at "my_db_server" (X.X.X.X), port 5432 failed: fe_sendauth: error sending password a
uthentication
```

I have debugged the code to confirm the password made it correctly to the connect call, also confirmed there is communication with the database server. I did not proceed with debugging postgres server to see why it rejected. Updating the requirements.txt to pull psycopg2-binary==2.9.1 recovered the situation.

I doubt the openssl version could be related as we do not use SSL over the connection, but perhaps it is used to encrypt password? FIPS is enabled so probably this can be related, I had to patch django with https://code.djangoproject.com/ticket/28401 to get around that.

It is obviously not a netbox problem, I suspect that there could be a change in requirements of psycopg2 that netbox should carry over and are not met in our system. I am wondering why netbox pulls the binary package, because gcc and dev libraries are listed as netbox dependency and I noticed that pip install also used them. So there should be no problem to build custom postgres python client also.

**Reply by candlerb** [reply]
ISTM that if you're in an environment that requires FIPS, then enabling MD5 is not a great idea.  Did you find Netbox requires MD5 for some other reason apart from Postgres login?  If MD5 is being used for postgres authentication, it's clearly being "used for security".

Fortunately, Postgres 10+ supports [scram-sha-256](https://www.postgresql.org/docs/10/auth-methods.html#AUTH-PASSWORD) which you can use instead of md5.

Netbox 3.1 will require a minimum of postgres 10, so at that point it would make sense to update the documentation to show this instead of md5 auth.  Of course you can switch now, as long as your postgres server is 10 or later.

**Reply by candlerb** [reply]
BTW, I just upgraded my home system to scram-sha-256 to ensure this works, and it does.  [Useful blog here](https://blog.crunchydata.com/blog/how-to-upgrade-postgresql-passwords-to-scram).

**Reply by amk1969** [reply]
The use of MD5 in django is for caching, not security. There I have it enabled. Could it happen that psycopg2-binary 2.9.2 would make use of a cipher/hash that is not allowed by FIPS while version 2.9.1 does not have any issue?

The server is new installed version 13.3 but I took some settings from former one so will have to check what is setup there. Thanks for the useful pointer. The server has also FIPS enabled so I would assume no one should be able to connect with MD5. Is it possible that postgres library before version 14 does not obey FIPS?

**Reply by candlerb** [reply]
> Could it happen that psycopg2-binary 2.9.2 would make use of a cipher/hash that is not allowed by FIPS while version 2.9.1 does not have any issue?

I was thinking the other way round: that your 2.9.2 *didn't* allow MD5 while your 2.9.1 does.  However, it's just speculation, since I don't know really what's different between 2.9.1 and 2.9.2 on your system. 2.9.2 works fine for me, with md5 auth as well as scram-sha-256.

I don't imagine the postgres library knows anything about FIPS. That is, I think it would be up to someone to release a FIPS-certified postgres client library, if there were such a thing.

**Reply by amk1969** [reply]
The difference between 2.9.1 and 2.9.2 on all systems is bump of Openssl version as well as postgres client version. I find it is possible that one of them would add FIPS support, leading to no MD5 anymore. Still I would expect something relevant logged except "error sending password". I will know more tomorrow once I have a look at the sql server.

**Comment by amk1969** [comment]
It is confirmed that the MD5 hashing at the server side was the culprit. After changing to scram-sha-256 as suggested by user candlerb
yesterday the connection works also with recent version of psycopg2. It is the postgres library version 14 inside psycopg2 that brought FIPS support to the client: https://www.postgresql.org/docs/14/release-14.html#id-1.11.6.6.5

Thanks for all help.
